restore dynamic IRQ API

tyrantlizard77 opened the pull request:
Zephyr used to support installing interrupt handlers at runtime, but this capability was removed as we could not find any use-cases for doing it and the x86 implementation isn't simple.

However, we'd like to support hypervisors like ACRN which present virtual peripherals that enumerate as PCI. We don't know the IRQ parameters at build time and so here we restore the original dynamic interrupt implementation with some enhancements.

Commit:
kernel: add dynamic interrupt API

In the past the capability to install interrupts at runtime was
removed due to lack of use-cases for Zephyr's intended targets.

Now we want to support hypervisor applications like ACRN where
virtual devices are presented to the kernel using PCI enumeration,
and the interrupt configuration is not known at build time.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
x86: add dynamic interrupt support

If dynamic interrupts are enabled, a set of trampoline stubs
are generated which transfer control to a common dynamic
interrupt handler function, which then looks up the proper
handler and parameter and then executes the interrupt.

Based on the prior x86 dynamic interrupt implementation which
was removed from the kernel some time ago, and adapted to
changes in the common interrupt handling code, build system,
and IDT generation tools.

An alternative approach could be to read the currently executing
vector out of the APIC, but this is a much slower operation.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
doc: add details about dynamic interrupts

Update interrupts.rst with information on how to use dynamic
interrupts. As they are used in the same way as IRQ_CONNECT(),
not much needs to be written.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
arch: common: add function for updating IRQ table

This will be called by arch-specific implementations of
_arch_irq_connect_dynamic()

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
arc: add _arch_irq_connect_dynamic()

Implements dynamic interrupts.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
arm: add _arch_irq_connect_dynamic()

Implements dynamic interrupts.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
nios2: add _arch_irq_connect_dynamic()

Implements dynamic interrupts.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
riscv32: add _arch_irq_connect_dynamic()

Implements dynamic interrupts.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
xtensa: add _arch_irq_connect_dynamic()

Implements dynamic interrupts.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented:
Adding a DNM label until I can do a full set of hardware testing.

pfalcon commented:
@ozhuraki should be happy too, as he raised this question in https://github.com/zephyrproject-rtos/zephyr/pull/9112

rgundi commented on arch/Kconfig:
```diff
@@ -202,6 +202,14 @@ menu "Interrupt Configuration"
 #
 # Interrupt related configs
 #
+config DYNAMIC_INTERRUPTS
```
Shouldn't this depend on "GEN_SW_ISR_TABLE"?

andrewboie replied:
no, this is a generic option. x86 doesn't use SW ISR table, it has a different implementation.

rgundi commented on include/linker/sections.h:
```diff
@@ -46,7 +46,15 @@
-#define SW_ISR_TABLE		.gnu.linkonce.sw_isr_table
+
+/* Place the table in either RAM or ROM depending on whether we allow
+ * IRQs to be configured at runtime
```
I couldn't quite figure out what .gnu.linkonce does and how .gnu.linkonce.d and .gnu.linkonce.r differ from each other. May be a little more verbosity is required here.

andrewboie replied:
Use of this has been historical for some time, but when I looked it seems it is undocumented in the 'ld' documentation. I had to go to the source code for binutils:

```
  /* As a GNU extension, if the name begins with .gnu.linkonce, we
     only link a single copy of the section.  This is used to support
     g++.  g++ will emit each template expansion in its own section.
     The symbols will be defined as weak, so that multiple definitions
     are permitted.  The GNU linker extension is to actually discard
     all but one of the sections.  */
```

I'm going to see if I can do the same thing with __attribute__((weak)) which is at least documented.

> how .gnu.linkonce.d and .gnu.linkonce.r differ from each other

These are just sections in the linker script, placed in data and rodata respectively.

andrewboie replied:
yeah I think I can do this with __weak. updated PR to follow.

rgundi reviewed (commented):
Looks good for all non-x86 arches. I leave the review of x86 to subject matter experts.

dbkinder commented on doc/kernel/other/interrupts.rst:
```diff
@@ -389,6 +412,13 @@ creates an _irq_to_interrupt_vector array which maps an IRQ line to its
 configured vector in the IDT. This is used at runtime by :c:macro:`IRQ_CONNECT`
 to program the IRQ-to-vector association in the interrupt controller.
 
+For dynamic interrupts, it is necessary for the build to generate some 4-byte
```
How about,

    For dynamic interrupts, the build must generate 4-byte ...

andrewboie replied:
@dbkinder can you re-visit?

ruuddw commented on doc/kernel/other/interrupts.rst:
```diff
@@ -218,6 +218,27 @@ The following code defines and enables an ISR.
+
+Dynamic interrupts require the :option:`CONFIG_DYNAMIC_INTERRUPTS` option to
+be enabled. Removing or re-configuring a dynamic interrupt is currently
+unsupported.
```
I don't see anything that prevents re-configuring by calling irq_connect_dymanic multiple times for the same irq, at least not for the generic non-x86 case.

ruuddw reviewed (approved):
one minor documentation remark (re-configure not supported?)

Commit:
interrupts: simplify position of sw ISR table

We now place the linker directives for the SW ISR table
in the common linker scripts, instead of repeating it
everywhere.

The table will be placed in RAM if dynamic interrupts are
enabled.

A dedicated section is used, as this data must not move
in between build phases.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
interrupts: use weak linkage instead of linkonce

.gnu.linkonce is an internal undocumented ld feature.
Just use __weak, which does the same thing we want.

This is only done for _sw_isr_table. _irq_vector_table
is left alone due to unwanted interactions between
__weak and the ld KEEP() directive.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

rgundi reviewed (approved):
non-x86 looks good to me.

andyross commented on arch/common/sw_isr_common.c:
```diff
@@ -0,0 +1,28 @@
+void z_isr_install(unsigned int irq, void (*routine)(void *), void *param)
+{
+	int table_idx = irq - CONFIG_GEN_IRQ_START_VECTOR;
+	int key = irq_lock();
```
Don't see why you need to lock this.  If the user is in the process of installing a handler for an interrupt that is unmasked, then they have already lost the game.

Along the same lines, it couldn't hurt to add an __ASSERT(!irq_is_enabled(irq), "...") here.

andrewboie replied:
done.

andyross commented on arch/common/sw_isr_common.c:
```diff
@@ -0,0 +1,28 @@
+ * Common code for arches that use software ISR tables (CONFIG_GEN_ISR_TABLES)
+ */
+
+#ifdef CONFIG_DYNAMIC_INTERRUPTS
```
Pedantry: you don't need to wrap the definition of non-statics with an #ifdef.  They will only enter the link if called, and the declaration is already wrapped.

andyross commented on arch/x86/core/intstub.S:
```diff
@@ -464,3 +465,62 @@ SECTION_FUNC(TEXT, _irq_sw_handler)
+	jmp _interrupt_enter
+
+/* Create all the dynamic IRQ stubs
+ *
```
Nice macro work, FWIW.

andyross replied:
Drive-by comment: what I did instead with x86_64 is to have every IDT entry point to a separate 5-byte CALL instruction, all of which entered the same handler (well, two: with error code and without error code, but one falls through to the other).  Then you can extract the vector in C code with just a subtract and shift of the return address pushed by the call.

Single-byte-per-entry overhead (and one call instruction) vs. this scheme, but works symmetrically across all interrupts and doesn't need special handling for "dynamic" ones.

andrewboie replied:
I think it might be beneficial to unify these implementations for x86 in the fullness of time, I have no particular attachment to this one.

andyross commented on arch/x86/core/irq_manage.c:
```diff
@@ -99,3 +99,255 @@ void _arch_isr_direct_footer(int swap)
+ * Each interrupt priority level  contains 16 vectors, and the prioritization
+ * of interrupts within a priority  level is determined by the vector number;
+ * the higher the vector number, the higher the priority within that priority
+ * level.
```
Don't think this sentence is true.  The APIC will only trigger an interrupt on priority change, so there's no notion of priority "within that priority level".

andrewboie replied:
Removed. This was some very old verbiage from the Viper/WR days.

andyross reviewed (approved):
Looks clean.  Some of the x86 complexity is a grave we dug for ourselves with all the static IDT stuff, but I don't see an obvious way to fix that without having to have it in RAM.

lpereira commented:
Unless you're planning on having device hotplug support, it would be nice if the table could be marked as read-only after the boot is complete.  It's a good security measure.  Linux has an interesting feature: [mark a variable as "ro_after_init", and the pages for those variables are then marked as such](https://lwn.net/Articles/676145/).  Maybe not possible on MPU devices due to the limited amount of regions, but for MMU devices with page tables, this is a nice-to-have feature.

andrewboie commented:
> Unless you're planning on having device hotplug support, it would be nice if the table could be marked as read-only after the boot is complete

On an MMU system it would have to have it's own dedicated 4k page for this, since we have no other data structures with these semantics.

Maybe I'm not quite understanding the value of this feature, if the attacker somehow gets the ability to write to system RAM, they can modify the page tables in RAM as they please and remove any RO protections anyway

Either way, not in this PR.
